src/plugins/intel_gpu/src/graph/impls/ocl_v2/moe/moe_3gemm_swiglu_opt.cpp

```diff
@@ -920,6 +1168,254 @@ class moe_3gemm_swiglu_opt_impl : public PrimitiveImplOCL {
+        auto routing_mem_ptr = scratch.topk_weights;
+        auto input_shape = instance.input_memory_ptr(static_cast<size_t>(MOE3GemmInputIndex::HIDDEN_STATES))->get_layout().get_shape();
+
+        _hidden_size = static_cast<int>(cur_moe->_config.hidden_size);
+        _intermediate_size = static_cast<int>(cur_moe->_config.inter_size);
+
+        auto rtp = static_cast<MoE3GemmRuntimeParams*>(m_rt_params.get());
+        const size_t subgroup_size = instance.get_impl_params()->get_device_info().arch >= gpu_arch::xe2 ? 32 : 16;
+        // const size_t max_work_group_size = instance.get_impl_params()->get_device_info().max_work_group_size;
+
+        event::ptr ret_event;
+        const auto& intermediates_memories = instance.get_intermediates_memories();
+        auto& stream = instance.get_network().get_stream();
+        auto num_total_experts = static_cast<int>(cur_moe->_config.num_expert);
+        int num_actually_used_experts = 0;
+
+        expert_mask_cpu expert_mask_cpu;
+        get_expert_mask_from_gpu(config, batch_mem_ptr, stream, expert_mask_cpu);
+
+        // step 1: generate 4 mask data for following kernel execution
+        // input: topk output, [token_len, expert_topk]
+        // output:
+        //   mask 0: token idx per expert, flat array of length token_len * expert_topk
+        //             (experts are laid out consecutively; use experts_info_start_idx + tokens_lens_per_expert to slice)
+        //   mask 1: token start offset idx in mask 0 for each activated expert, shape = [activated_expert_num]
+        //   mask 2: token len for each activated expert, shape = [activated_expert_num]
+        //   mask 3: expert id, shape = [activated_expert_num]
+        {
+            auto token_size = input_shape[0];
+            auto max_topk = static_cast<int>(cur_moe->_config.top_k);
+            std::vector<int32_t> tokens_per_expert_cpu(token_size * max_topk, -1);
+            std::vector<int32_t> tokens_lens_per_expert_cpu(num_total_experts, -1);
+            std::vector<int32_t> experts_info_start_idx_cpu(num_total_experts, -1);
+            std::vector<int32_t> experts_id_cpu(num_total_experts, -1);
+
+            int tokens_per_expert_iter = 0;
+            int experts_id_iter = 0;
+            for (int expert_idx = 0; expert_idx < num_total_experts; expert_idx++) {
+                if (!expert_mask_cpu.batch[expert_idx].empty()) {
+                    experts_info_start_idx_cpu[experts_id_iter] = tokens_per_expert_iter;
+                    experts_id_cpu[experts_id_iter] = expert_idx;
+                    tokens_lens_per_expert_cpu[experts_id_iter++] = static_cast<int32_t>(expert_mask_cpu.batch[expert_idx].size());
+                    num_actually_used_experts++;
+                    for (auto t : expert_mask_cpu.batch[expert_idx]) {
+                        tokens_per_expert_cpu[tokens_per_expert_iter++] = t;
+                    }
+                }
+            }
```

Comment:
Can we use a GPU kernel to get the required expert mask info? Then no need to interrupt gpu execution here.